pyquil/quilbase.py

```diff
@@ -1292,48 +1331,112 @@ def out(self) -> str:
+    def __new__(
+        cls,
         name: str,
-        parameters: List[ParameterDesignator],
-        qubits: List[Union[Qubit, FormalArgument]],
-        instrs: List[AbstractInstruction],
+        parameters: Iterable[ParameterDesignator],
+        qubits: Iterable[Union[Qubit, FormalArgument]],
+        instrs: Iterable[AbstractInstruction],
+        modifiers: Iterable[quil_rs.GateModifier] = [],
     ):
-        self.name = name
-        self.parameters = parameters
-        self.qubits = qubits
-        self.instrs = instrs
+        return super().__new__(
+            cls,
+            name,
+            _convert_to_rs_expressions(parameters),
+            _convert_to_rs_qubits(qubits),
+            _convert_to_rs_instructions(instrs),
+            modifiers,
+        )
+
+    @classmethod
+    def _from_rs_calibration(cls, calibration: quil_rs.Calibration) -> "DefCalibration":
+        return cls(
+            calibration.name,
+            calibration.parameters,
+            calibration.qubits,
+            calibration.instructions,
+            calibration.modifiers,
+        )
+
+    @property
+    def parameters(self):
+        return _convert_to_py_parameters(super().parameters)
+
+    @parameters.setter
+    def parameters(self, parameters: Iterable[ParameterDesignator]):
+        quil_rs.Calibration.parameters.__set__(self, _convert_to_rs_expressions(parameters))
+
+    @property
+    def qubits(self):
+        return _convert_to_py_qubits(super().qubits)
+
+    @qubits.setter
+    def qubits(self, qubits: Iterable[QubitDesignator]):
+        quil_rs.Calibration.qubits.__set__(self, _convert_to_rs_qubits(qubits))
+
+    @property
+    def instrs(self):
+        return _convert_to_py_instructions(super().instructions)
+
+    @instrs.setter
+    def instrs(self, instrs: Iterable[AbstractInstruction]):
+        quil_rs.Calibration.instructions.__set__(self, _convert_to_rs_instructions(instrs))
 
     def out(self) -> str:
-        ret = f"DEFCAL {self.name}"
-        if len(self.parameters) > 0:
-            ret += _format_params(self.parameters)
-        ret += " " + _format_qubits_str(self.qubits) + ":\n"
-        for instr in self.instrs:
-            ret += f"    {instr.out()}\n"
-        return ret
+        return str(self)
 
 
-class DefMeasureCalibration(AbstractInstruction):
-    def __init__(
-        self,
+class DefMeasureCalibration(quil_rs.MeasureCalibrationDefinition, AbstractInstruction):
+    def __new__(
+        cls,
         qubit: Union[Qubit, FormalArgument],
         memory_reference: Optional[MemoryReference],
         instrs: List[AbstractInstruction],
-    ):
-        self.qubit = qubit
-        self.memory_reference = memory_reference
-        self.instrs = instrs
+    ) -> "DefMeasureCalibration":
+        return super().__new__(
+            cls,
+            _convert_to_rs_qubit(qubit),
+            "" if memory_reference is None else str(memory_reference),
```

Comment:
Yeah this is a bit of a mess - 

* the [spec](https://github.com/quil-lang/quil/blob/master/rfcs/analog/spec_changes.md#defining-calibrations) says a parameter: 

```
GateModifier :: CONTROLLED | DAGGER | FORKED
CalibrationDefinition :: DEFCAL OpModifier* Name ( Parameter+ ) Qubit+ : Instruction+
MeasureCalibrationDefinition :: DEFCAL Name Qubit? Parameter : Instruction+
```

* The example shows a prefixed parameter:

```
# Measurement and classification
DEFCAL MEASURE 0 %dest:
    DECLARE iq REAL[2]
    CAPTURE 0 "out" flat(1e-6, 2+3i) iq
    LT %dest iq[0] 0.5 # thresholding
```

* But live, production calibrations use a non-prefixed `addr` (because that's using the pyQuil code):

```
DEFCAL MEASURE 32 addr:
    FENCE 32
    DECLARE q32_unclassified REAL[2]
    # ...
    NONBLOCKING CAPTURE 32 "ro_rx" boxcar_kernel(...) addr
    # ...
```

So I think here we have several bugs cancelling out:

* `quil_rs` should prefix the parameter with a `%` but it doesn't, and so successfully parses and writes something that looks like a memory reference
* `pyQuil should consider this a variable rather than a memory reference, per the spec
* Really, the spec should consider this a memory reference

It's not a memory reference per the spec, even though that's how it's always used within the body of the calibration. I think it _should_ be, and that's what @notmgsk was likely thinking as well when he wrote that part of the pyQuil parser.

In any case, it should not be optional - measurement needs a target. The spec is fuzzy on that but that's something I'd like to be clarified as well.

Reply:
Thank you that all makes sense. So what should we do here? I'm inclined to make both `DefMeasureCalibration` and `MeasureCalibrationDefinition` require a `MemoryReference` as a destination. That's a breaking change, but it avoids this messiness and aligns with how we expect this to be used.

Reply:
Sure, that works, we'll sort out the spec later and come back to this if needed.

Reply:
I've made the MemoryReference required on the pyQuil side and labeled the commit as a `BREAKING CHANGE` to keep track of it.